### ISSUE TITLE
Constellation hardening metabug

### ISSUE THREAD

**Issue by asajeffrey** [servo/servo#10124]
We should harden the constellation so that constellation panics are vanishingly rare.

Currently, we have a number of panics that take down the constellation. This should not happen! The constellation code already handles failure messages by replacing the pipeline with about:failure, this should happen for all panics.

Related issues:

* #10017 (has some relevant discussion)
* #9828 
* #9724
* #9387
* #9202 
* #8989
* #8815
* #7826 

Also see the discussion on PR #10082.

Notes on hardening: https://public.etherpad-mozilla.org/p/servo-threads

**Comment by asajeffrey** [comment]
There are four sources of panic in constellation.rs:

1. Lookup failure in the pipeline/frame maps (addressed by PR #10082)
2. Creating new processes
3. Creating new channels
4. Sending/receiving on those channels

The trickiest to deal with is (4). There's already code to handle failure messages in constellation.rs, should we treat errors on send as the same as receiving a failure message?

**Comment by nox** [comment]
> The trickiest to deal with is (4). There's already code to handle failure messages in constellation.rs, should we treat errors on send as the same as receiving a failure message?

The Erlang user in me says yes. Constellation should never die.

**Comment by asajeffrey** [comment]
Testing the hardened constellation requires some way of injecting panics, see #10144.

**Comment by asajeffrey** [comment]
Some IRC discussion http://logs.glob.uno/?c=mozilla%23servo#c390119 about double-panics.

TL;DR: #6462 causes double-panic, when SM panics and leaves the GC roots in an unsafe state, and this double-panic aborts the process. We'd need multi-process execution to isolate pipelines (in particular, put the constellation in a separate process).

**Comment by jdm** [comment]
The constellation already lives in the parent process.

**Comment by asajeffrey** [comment]
I started on a patch to randomly close pipelines: https://github.com/asajeffrey/servo/commit/a9fb350d5212864ea544516a08a7d5abeae09f0f

**Comment by asajeffrey** [comment]
The patch to randomly close pipelines is now PR #10179.

**Comment by asajeffrey** [comment]
In order to run wpt-test with random pipeline closure, we need to be able to pass command line arguments in to servo via wptrunner. This is the issue: https://github.com/w3c/wptrunner/issues/176 and this is the PR: https://github.com/w3c/wptrunner/pull/177

**Comment by asajeffrey** [comment]
I submitted PR #10242, which removes constellation panics caused by `chan.send(msg).unwrap()`.

**Comment by asajeffrey** [comment]
Over in https://github.com/servo/servo/pull/10242#issuecomment-202949609, @emilio points out that we should use `warn!` rather than `debug!` for logging unexpected cases.

**Comment by asajeffrey** [comment]
PR #10242 merged.

**Comment by asajeffrey** [comment]
I submitted PR #10295, which fixes some miscellaneous constellation panics. At this point, I think the only remaining panics are unavoidable:
* Failure of message receive (can only happen in cases where the channel invariants have been vioated).
* Failure of ipc-channel or process creation (requires a broader low-resource mechanism).

**Comment by pcwalton** [comment]
Why do recv failures always indicate bugs? Couldn't a crashed pipeline (in multiprocess mode, or just a panicking one) cause arbitrary recv failures in the constellation?

**Comment by asajeffrey** [comment]
@paulrouget in #10334 points out that we should notify browser.html when a pipeline panics.

**Comment by asajeffrey** [comment]
PRs #10082 and #10295 landed.

**Comment by asajeffrey** [comment]
PR #10294 ran into homu trouble, and is now PR #10343.

**Comment by asajeffrey** [comment]
PR #10345 addresses some coding style issues raised by @nox, about using `unwrap_or_else` for side-effects, rather than `if let`.

**Comment by asajeffrey** [comment]
#10446 puts a test in to travis to ensure against regressions.

**Comment by asajeffrey** [comment]
#10899 adds a test for constellation hardening. If we're feeling brave we could gate on it.

**Comment by asajeffrey** [comment]
Once #10899 lands, we can close this metabug, woo hoo!

**Comment by paulrouget** [comment]
#10899 has landed.
